### ISSUE TITLE
io_device unregister data race

### ISSUE THREAD

**Issue by darsto** [spdk/spdk#278]
I managed to hit an assert in io_channel.c:374
https://github.com/spdk/spdk/blob/a8d3ac0cf4b812d7cda5d0abb4dfe1cdc999fb56/lib/util/io_channel.c#L374
using basic bdevio test with a single virtio bdev.

The gdb trace below shows that that one thread has already unregistered the io_device and is calling post-unregister virtio routine while the other thread hits an assert within _put_io_channel. I believe at the time when our assert is first encountered, the `spdk_io_device_unregister` is still in progress.

A brief look at the code shows the following data race:

`_spdk_io_device_attempt_free (called by _put_io_channel)`:
https://github.com/spdk/spdk/blob/a8d3ac0cf4b812d7cda5d0abb4dfe1cdc999fb56/lib/util/io_channel.c#L369-L376

`spdk_io_device_unregister`:
https://github.com/spdk/spdk/blob/a8d3ac0cf4b812d7cda5d0abb4dfe1cdc999fb56/lib/util/io_channel.c#L403-L407

`dev->unregister_cb` must have been already set, but `dev->unregister_thread` store is still in progress.

Shouldn't all this code be protected by `g_devlist_mutex`?

Gdb trace:
```
(gdb) i thr
  Id   Target Id         Frame 
* 1    Thread 9173.9173 "reactor_0" 0x000055555564643c in _finish_unregister (arg=0x6070000068b0) at io_channel.c:336
  2    Thread 9173.9195 "reactor_2" __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:51
  3    Thread 9173.9191 "rte_mp_handle" 0x00007ffff6004df0 in __libc_recvmsg (fd=7, msg=0x7ffff2bfba30, flags=0) at ../sysdeps/unix/sysv/linux/recvmsg.c:28
  4    Thread 9173.9192 "vfio-sync" 0x00007ffff60043ed in accept () at ../sysdeps/unix/syscall-template.S:84
  5    Thread 9173.9193 "eal-intr-thread" 0x00007ffff5d3e0f3 in epoll_wait () at ../sysdeps/unix/syscall-template.S:84
  6    Thread 9173.9194 "reactor_1" _spdk_event_queue_run_batch (reactor=0x61800000fd80) at reactor.c:175

(gdb) thr 2
(gdb) bt
#3  0x00007ffff5c80ee2 in __GI___assert_fail (assertion=0x5555556c2a20 "dev->unregister_thread != NULL", file=0x5555556c2780 "io_channel.c", line=374, function=0x5555556c2e40 <__PRETTY_FUNCTION__.7777> "_spdk_io_device_attempt_free") at assert.c:101
#4  0x00005555556467ed in _spdk_io_device_attempt_free (dev=0x6070000068b0, ch=0x0) at io_channel.c:374
#5  0x0000555555647429 in _spdk_put_io_channel (arg=0x60700000a740) at io_channel.c:501
#6  0x000055555563a58b in _spdk_reactor_msg_passed (arg1=0x555555647232 <_spdk_put_io_channel>, arg2=0x60700000a740) at reactor.c:210

(gdb) thr 1
(gdb) bt
[...]
#7  0x0000555555619838 in virtio_dev_stop (dev=0x61300000d400) at virtio.c:682
#8  0x0000555555613558 in virtio_blk_dev_unregister_cb (io_device=0x61300000d400) at bdev_virtio_blk.c:228
#9  0x000055555564643c in _finish_unregister (arg=0x6070000068b0) at io_channel.c:336
#10 0x000055555563a58b in _spdk_reactor_msg_passed (arg1=0x5555556463ce <_finish_unregister>, arg2=0x6070000068b0) at reactor.c:210
```

**Comment by dverkamp-intel** [comment]
Can you check if this is fixed by commit 979d9997c9a800c85afe2e8d0d384ede62158101?

**Comment by darsto** [comment]
I think it only made it worse.

Please see here - https://review.gerrithub.io/c/408134/

2018-04-17 19:33 GMT+02:00 Daniel Verkamp <notifications@github.com>:
